### ISSUE TITLE
FSDP enable users to seamlessly switch between DDP, ZeRO-1, ZeRO-2 and FSDP flavors of data parallelism

### ISSUE THREAD

**Issue by GuWei007** [pytorch/pytorch#102532]
### 🚀 The feature, motivation and pitch

https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/
DeepSpeed is hard to use,  FSDP is easy to use for LLM, but FSDP do not support ZeRO-1, ZeRO-2,  we suggest FSDP enable users to seamlessly switch between DDP, ZeRO-1, ZeRO-2 and FSDP flavors of data parallelism

cc @zhaojuanmao @mrshenli @rohan-varma @awgu

**Comment by LicharYuan** [comment]
> * ZeRO-2 corresponds to [`ShardingStrategy.SHARD_GRAD_OP`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/api.py#L43).
> * ZeRO-1 corresponds to `SHARD_GRAD_OP` with gradient accumulation with [`no_sync()`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/fully_sharded_data_parallel.py#L965). Otherwise, there is no point to use ZeRO-1 since ZeRO-2 is strictly better.
> * FSDP's version of DDP is [`ShardingStrategy.NO_SHARD`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/api.py#L50).
> 
> These are all configured by passing to the [`sharding_strategy`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/fully_sharded_data_parallel.py#L226) argument.
> 
> Could you also explain more on why DeepSpeed is hard to use? We are thinking about API design lately, and I would love to get more feedback to understand the pain points.

Will FSDP support ZeRO-3 (parameters-partition)?

**Comment by awgu** [comment]
@LicharYuan ZeRO-3 is the default for FSDP (`FULL_SHARD`) :)

**Comment by LicharYuan** [comment]
> @LicharYuan ZeRO-3 is the default for FSDP (`FULL_SHARD`) :)

lol, Thx!

**Comment by Mddct** [comment]
zero2 

> * ZeRO-2 corresponds to [`ShardingStrategy.SHARD_GRAD_OP`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/api.py#L43).
> * ZeRO-1 corresponds to `SHARD_GRAD_OP` with gradient accumulation with [`no_sync()`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/fully_sharded_data_parallel.py#L965). Otherwise, there is no point to use ZeRO-1 since ZeRO-2 is strictly better.
> * FSDP's version of DDP is [`ShardingStrategy.NO_SHARD`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/api.py#L50).
> 
> These are all configured by passing to the [`sharding_strategy`](https://github.com/pytorch/pytorch/blob/0f1621df1a0a73956c7ce4e2f72f069e610e0137/torch/distributed/fsdp/fully_sharded_data_parallel.py#L226) argument.
> 
> Could you also explain more on why DeepSpeed is hard to use? We are thinking about API design lately, and I would love to get more feedback to understand the pain points.

but zero2 is slow than zero1?

**Comment by awgu** [comment]
@Mddct I think it depends on what you consider ZeRO-1. Could you explain why ZeRO-2 is slower than ZeRO-1?

**Comment by Mddct** [comment]
> @Mddct I think it depends on what you consider ZeRO-1. Could you explain why ZeRO-2 is slower than ZeRO-1?

I think zero2 has the overhead of gradient shard merging

**Comment by awgu** [comment]
I am not familiar with "gradient shard merging". Could you explain what that is?

**Comment by Fanoid** [comment]
> I am not familiar with "gradient shard merging". Could you explain what that is?

I think what @Mddct means is ZeRO-2 have to do gradient sync (reduce-scatter) after every micro-batch when with gradient accumulation, but ZeRO-1 only sync once.

**Comment by awgu** [comment]
@Fanoid Got it! For FSDP, there is a `no_sync` option that allows for ZeRO-1 semantics when microbatching with gradient accumulation.

**Comment by withoutaword** [comment]
hi @awgu  I am using FSDP with trainer and fsdp_config as below . I would like to try using ZeRO1 , where should I config no_syc, thanks ! 

{
  "compute_environment": "LOCAL_MACHINE",
  "debug": false,
  "distributed_type": "FSDP",
  "downcast_bf16": "no",
  "fsdp_config": {
    "fsdp_auto_wrap_policy": "TRANSFORMER_BASED_WRAP",
    "fsdp_backward_prefetch": "BACKWARD_PRE",
    "fsdp_cpu_ram_efficient_loading": true,
    "fsdp_forward_prefetch": false,
    "fsdp_offload_params": false,
    "fsdp_sharding_strategy": 2,
    "fsdp_state_dict_type": "FULL_STATE_DICT",
    "fsdp_sync_module_states": true,
    "fsdp_use_orig_params": false
  },
  "main_training_function": "main",
  "mixed_precision": "no",
  "rdzv_backend": "static",
  "same_network": true,
  "tpu_env": [],
  "tpu_use_cluster": false,
  "tpu_use_sudo": false,
  "num_processes": 4,
  "use_cpu": false
}

**Comment by awgu** [comment]
@withoutaword sorry, I am not familiar with how the trainer is setup to know how to pass `no_sync` :/
